base/resolver_node_lookup.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package base
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const lookupTimeout = time.Second * 3
+
+// nodeLookupResolver implements Resolver.
+// It queries http(s)://<address>/_status/nodes and extracts the first node's address.
+// This is useful for http load balancers which will not forward RPC.
+// It is never exhausted.
+type nodeLookupResolver struct {
+	context   *Context
+	typ       string
+	addr      string
+	exhausted bool
+	// We need our own client so that we may specify timeouts.
+	httpClient *http.Client
+}
+
+// Type returns the resolver type.
+func (nl *nodeLookupResolver) Type() string { return nl.typ }
+
+// Addr returns the resolver address.
+func (nl *nodeLookupResolver) Addr() string { return nl.addr }
+
+// GetAddress returns a net.Addr or error.
+// Upon errors, we set exhausted=true, then flip it back when called again.
+func (nl *nodeLookupResolver) GetAddress() (net.Addr, error) {
+	if nl.exhausted {
+		nl.exhausted = false
+		return nil, util.Errorf("skipping temporarily-exhausted resolved")
+	}
+
+	if nl.httpClient == nil {
+		tlsConfig, err := nl.context.GetClientTLSConfig()
+		if err != nil {
+			return nil, err
+		}
+		nl.httpClient = &http.Client{
+			Transport: &http.Transport{TLSClientConfig: tlsConfig},
+			Timeout:   lookupTimeout,
+		}
+	}
+
+	nl.exhausted = true
+	// TODO(marc): put common URIs in base and reuse everywhere.
+	url := fmt.Sprintf("%s://%s/_status/nodes", nl.context.RequestScheme(), nl.addr)
```

Comment:
This doesn't feel like the correct endpoint for this. Correct me if I'm wrong, but you're running into an issue where you need to get the address of a node in the cluster via HTTP because the load balancer play nicely with HTTP forwarding. You then use the address to directly connect to a gossip node.

What you actually want to do is improve the `/_status/local` endpoint by adding the address of the node to the json which is currently returned (it currently just has the build info). You could then just make that HTTP request repeatedly and get a random sampling of the external IP addresses of the nodes in the cluster.

I'll send you a PR which provides the address there.

Going to `/_status/nodes` is problematic because most nodes won't be able to write updated information there without first using gossip to connect to the cluster, so they can get first range metadata, etc.